File: src/main/java/com/team573/gongguri/domain/groupPurchase/controller/GroupPurchaseController.java
package com.team573.gongguri.domain.groupPurchase.controller;

import com.team573.gongguri.domain.chat.repository.ChatRoomRepository;
import com.team573.gongguri.domain.groupPurchase.dto.GroupPurchaseRequestDto;
import com.team573.gongguri.domain.groupPurchase.dto.GroupPurchaseResponseDto;
import com.team573.gongguri.domain.groupPurchase.dto.GroupPurchaseWithChatResponseDto;
import com.team573.gongguri.domain.groupPurchase.entity.ProgressStatus;
import com.team573.gongguri.domain.groupPurchase.service.GroupPurchaseService;
import com.team573.gongguri.domain.member.repository.MemberRepository;
import com.team573.gongguri.domain.member.repository.UnivRepository;
import com.team573.gongguri.global.security.CustomUserDetails;
import java.util.ArrayList;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/group-purchases")
@RequiredArgsConstructor
@Slf4j
public class GroupPurchaseController {
    private final GroupPurchaseService service;
    private final MemberRepository memberRepository;
    private final UnivRepository univRepository;
    private final ChatRoomRepository chatRoomRepository;
    private final GroupPurchaseService groupPurchaseService;

    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<GroupPurchaseResponseDto> add(@RequestBody GroupPurchaseRequestDto dto) {
        log.info("[GroupPurchaseController] JSON 방식 게시글 작성 요청 수신");
        GroupPurchaseResponseDto createdDto = service.add(dto);
        return ResponseEntity.status(HttpStatus.CREATED).body(createdDto);
    }

    @GetMapping("/{id}")
    public ResponseEntity<GroupPurchaseResponseDto> get(@PathVariable Long id) {
        return ResponseEntity.ok(service.get(id));
    }

    @GetMapping
    public ResponseEntity<List<GroupPurchaseResponseDto>> getAll() {
        return ResponseEntity.ok(service.getAll());
    }

    @PutMapping("/{id}")
    public ResponseEntity<GroupPurchaseResponseDto> update(@PathVariable Long id,
                                                           @RequestBody GroupPurchaseRequestDto dto) {

        return ResponseEntity.ok(service.update(id, dto));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        service.delete(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/chat")
    public ResponseEntity<List<GroupPurchaseWithChatResponseDto>> getWithChat(
        @RequestParam(required = false, name = "cursor") Long cursorGroupPurchaseId,
        @RequestParam(required = false) String progressStatus,
        @RequestParam(defaultValue = "10") int size,
        @AuthenticationPrincipal CustomUserDetails customUserDetails
    ) {
        List<ProgressStatus> progressStatuses = new ArrayList<>();

        if (progressStatus != null) {
            if (progressStatus.equals("RECRUITING")) {
                progressStatuses.add(ProgressStatus.RECRUITING);
                progressStatuses.add(ProgressStatus.CLOSED);
            } else if (progressStatus.equals("COMPLETED")) {
                progressStatuses.add(ProgressStatus.COMPLETED);
            }
        }

        List<GroupPurchaseWithChatResponseDto> withMessages
            = groupPurchaseService.getWithMessage(size, cursorGroupPurchaseId, progressStatuses, customUserDetails.getMemberId());
        return ResponseEntity.ok(withMessages);
    }
}

File: src/main/java/com/team573/gongguri/domain/grouppurchase/controller/GroupPurchaseController.kt
package com.team573.gongguri.domain.grouppurchase.controller

import com.team573.gongguri.domain.grouppurchase.dto.*
import com.team573.gongguri.domain.grouppurchase.entity.PurchaseFilter
import com.team573.gongguri.domain.grouppurchase.service.GroupPurchaseParticipantService
import com.team573.gongguri.domain.grouppurchase.service.GroupPurchaseService
import com.team573.gongguri.global.security.CustomUserDetails
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/group-purchases")
class GroupPurchaseController(
    private val groupPurchaseService: GroupPurchaseService,
    private val groupPurchaseParticipantService: GroupPurchaseParticipantService
) {

    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun add(
        @RequestBody dto: GroupPurchaseRequestDto,
        @AuthenticationPrincipal customUserDetails: CustomUserDetails
    ): ResponseEntity<GroupPurchaseCreateResponseDto> {
        log.info("[GroupPurchaseController] JSON 방식 게시글 작성 요청 수신")
        val memberId = customUserDetails.memberId
        val createdDto = groupPurchaseService.add(dto, memberId)
        return ResponseEntity.status(HttpStatus.CREATED).body(createdDto)
    }

    @GetMapping("/{id}")
    operator fun get(
        @PathVariable id: Long,
        @AuthenticationPrincipal userDetails: CustomUserDetails
    ): ResponseEntity<GroupPurchaseDetailResponseDto> {
        val memberId = userDetails.memberId
        return ResponseEntity.ok(groupPurchaseService[id, memberId])
    }


    @PutMapping("/{id}")
    fun update(
        @PathVariable id: Long,
        @RequestBody dto: GroupPurchaseRequestDto
    ): ResponseEntity<GroupPurchaseUpdateResponseDto> {
        return ResponseEntity.ok(groupPurchaseService.update(id, dto))
    }


    @DeleteMapping("/{id}")
    fun delete(
        @PathVariable id: Long,
        @AuthenticationPrincipal userDetails: CustomUserDetails
    ): ResponseEntity<Void> {
        val memberId = userDetails.memberId
        groupPurchaseService.delete(id, memberId)
        return ResponseEntity.noContent().build()
    }

    @PostMapping("/{id}/join")
    fun join(
        @PathVariable id: Long,
        @AuthenticationPrincipal customUserDetails: CustomUserDetails
    ): ResponseEntity<Void> {
        val memberId = customUserDetails.memberId
        groupPurchaseService.join(id, memberId)
        log.info("member joined: {}", memberId)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/chat")
    fun getWithChat(
        @RequestParam(required = false, name = "cursor") cursorGroupPurchaseId: Long?,
        @RequestParam(required = false, defaultValue = "ALL", name = "progressStatus") purchaseFilter: PurchaseFilter,
        @RequestParam(defaultValue = "10") size: Int,
        @AuthenticationPrincipal customUserDetails: CustomUserDetails
    ): ResponseEntity<List<GroupPurchaseWithChatResponseDto>> {
        val withMessages = groupPurchaseService.getWithMessage(
            size,
            cursorGroupPurchaseId,
            purchaseFilter,
            customUserDetails.memberId
        )
        return ResponseEntity.ok(withMessages)
    }

    @PatchMapping("/{groupPurchaseId}/participants/{participantsId}/confirm")
    fun confirmDeposit(
        @PathVariable groupPurchaseId: Long,
        @PathVariable participantsId: Long,
        @AuthenticationPrincipal customUserDetails: CustomUserDetails
    ): ResponseEntity<Void> {
        groupPurchaseParticipantService.confirmDeposit(groupPurchaseId, participantsId, customUserDetails.memberId)
        return ResponseEntity.noContent().build()
    }

    @PatchMapping("/{groupPurchaseId}/participants/{participantsId}/cancel")
    fun cancelParticipantStatus(
        @PathVariable groupPurchaseId: Long,
        @PathVariable participantsId: Long,
        @AuthenticationPrincipal customUserDetails: CustomUserDetails
    ): ResponseEntity<Void> {
        groupPurchaseParticipantService.cancelParticipation(
            groupPurchaseId,
            participantsId,
            customUserDetails.memberId
        )
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/{groupPurchaseId}/participants")
    fun getParticipants(
        @PathVariable groupPurchaseId: Long,
        @RequestParam(required = false, name = "cursor") cursorParticipantId: Long?,
        @RequestParam(required = false) deposit: Boolean?,
        @RequestParam(defaultValue = "10") size: Int,
        @AuthenticationPrincipal customUserDetails: CustomUserDetails
    ): ResponseEntity<List<GroupPurchaseParticipantResponseDto>> {
        val participants = groupPurchaseParticipantService.getParticipants(
            groupPurchaseId, cursorParticipantId, deposit, customUserDetails.memberId, size
        )

        return ResponseEntity.ok(participants)
    }

    @GetMapping
    fun getAllByCursor(
        @RequestParam(required = false, name = "cursor") cursorGroupPurchaseId: Long?,
        @RequestParam(required = false, defaultValue = "ALL", name = "purchaseFilter") purchaseFilter: PurchaseFilter,
        @RequestParam(defaultValue = "10") size: Int
    ): ResponseEntity<List<GroupPurchaseListResponseDto>> {
        val progressStatuses = purchaseFilter.toStatuses()

        val groupPurchases = groupPurchaseService.getAllByCursor(
            cursorGroupPurchaseId,
            progressStatuses,
            size
        )

        return ResponseEntity.ok(groupPurchases)
    }
    companion object {
        private val log: Logger = LoggerFactory.getLogger(GroupPurchaseController::class.java)
    }
}

